File: lib/to_mdx/parsers/open_document/table/table_cell.rb
# frozen_string_literal: true

require_relative "../base"
require_relative "../mixins/parent_traversal"

module ToMdx
  module Parsers
    module Table
      class TableCell < Base
        def initialize(element = nil, context = {})
          super(element, context)

          parent_row = Mixins::ParentTraversal.find_parent(element, "table:table-row")
          @style = context[:styles].use(element&.attributes&.[]("table:style-name"))
          @default_style = context[:styles].use(parent_row&.attributes&.[]("table:default-cell-style-name"))
        end

        def to_mdx
          effective_style = @style.empty? ? @default_style : @style
          style_class = effective_style.empty? ? "" : " className=\"#{effective_style}\""
          "<td#{style_class}>#{children.map(&:to_mdx).join(" ")}</td>"
        end

        def to_s
          children.map(&:to_s).join(" | ")
        end

        register_for "table:table-cell"
      end
    end
  end
end

File: src/parsers/open-document/table/table-cell.ts
import dedent from "dedent-js";
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class TableCell extends BaseElement {
  private style: Style;
  private defaultStyle: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);

    this.style = context.styles.use(this.attr("table:style-name"));
    this.defaultStyle = context.styles.use(context.defaultCellStyleName);
  }

  toMdx(): string {
    const effectiveStyle = this.style.empty() ? this.defaultStyle : this.style;
    const noVPaddingStyle = effectiveStyle.without("padding-top", "padding-bottom");
    const styleClass = noVPaddingStyle.empty() ? "" : ` className="${noVPaddingStyle}"`;
    return dedent`
      <td${styleClass}>
        ${this.children.map((child) => child.toMdx()).join(" ")}
      </td>
    `;
  }

  toString(): string {
    return this.children.map((child) => child.toString()).join(" | ");
  }

  static {
    BaseElement.registerFor.call(this, "table:table-cell");
  }
}
